code/modules/vehicles/mecha/combat/justice.dm

```diff
@@ -0,0 +1,455 @@
+			if(stealth_action)
+				if(stealth_action.on)
+					stealth_action.Trigger()
+		button_icon_state = "mech_charge_on"
+		RegisterSignal(chassis, COMSIG_MECHA_MELEE_CLICK, PROC_REF(click_try_charge))
+	if(!on)
+		button_icon_state = "mech_charge_off"
+		UnregisterSignal(chassis, COMSIG_MECHA_MELEE_CLICK)
+	build_all_button_icons()
+
+///Called when mech attacks with charge attack enabled.
+/datum/action/vehicle/sealed/mecha/charge_attack/proc/click_try_charge(datum/source, mob/living/pilot, atom/target, on_cooldown, is_adjacent)
+	SIGNAL_HANDLER
+
+	var/turf = get_turf(target)
+	if(!on)
+		UnregisterSignal(chassis, COMSIG_MECHA_MELEE_CLICK)
+		return FALSE
+	if(!turf)
+		pilot.balloon_alert(pilot, "invalid direction!")
+		return FALSE
+	if(!charge)
+		pilot.balloon_alert(pilot, "on recharge!")
+		return FALSE
+	else
+		if(charge_attack(pilot, turf))
+			return TRUE
+	return FALSE
+
+/**
+ * ## charge_attack
+ *
+ * Brings mech out of invisibility.
+ * Deal everyone in range 3x3 35 damage and 25 chanse to cut off limb.
+ * Arguments:
+ * * charger - occupant inside mech.
+ * * target - occupant inside mech.
+ */
+/datum/action/vehicle/sealed/mecha/charge_attack/proc/charge_attack(mob/living/charger, turf/target)
+	/// Remember where we start charge
+	var/turf/start_charge_here = get_turf(charger)
+	/// Calculate charge range betwen range where we click and max charge range
+	var/charge_range = min(get_dist_euclidian(start_charge_here, target), max_charge_range)
+	/// Final result where where we start charge and charge range
+	var/turf/but_we_gonna_here = get_ranged_target_turf(start_charge_here, get_dir(start_charge_here, target), floor(charge_range))
+	/// Where mech is
+	var/turf/here_we_go = start_charge_here
+	for(var/turf/line_turf in get_line(get_step(start_charge_here, get_dir(start_charge_here, target)), but_we_gonna_here))
+		if(get_turf(charger) == get_turf(line_turf))
+			continue
+		if(isclosedturf(line_turf))
+			if(isindestructiblewall(line_turf))
+				break
+			if(istype(line_turf, /turf/closed/wall/r_wall))
+				line_turf.atom_destruction(MELEE)
+			if(istype(line_turf, /turf/closed/wall))
+				line_turf.atom_destruction(MELEE)
+		for(var/obj/break_in as anything in line_turf.contents)
+			if(istype(break_in, /obj/machinery/power/supermatter_crystal))
+				var/obj/machinery/power/supermatter_crystal/funny_crystal = break_in
+				funny_crystal.Bumped(chassis)
+				break
+			if(istype(break_in, /obj/machinery/gravity_generator))
+				continue
+			if(istype(break_in, /obj/machinery/atmospherics/pipe))
+				continue
+			if(istype(break_in, /obj/structure/disposalpipe))
+				continue
+			if(istype(break_in, /obj/structure/cable))
+				continue
+			if(istype(break_in, /obj/machinery) || istype(break_in, /obj/structure))
+				break_in.atom_destruction(MELEE)
+				continue
+		for(var/mob/living/somthing_living as anything in line_turf.contents)
```

Comment:
Why are you looping the same list twice?

Reply:
for mob and for obj